File: apps/burner-mac/Tests/FlagshipBurnerTests/VMResourcePlanTests.swift
import XCTest
@testable import FlagshipBurnerCore

/// Locks the resource-cap math from docs/desktop-vm-appliance.md: each VM +
/// the full data stack is several GB of RAM; single VM is the default; the
/// total is capped by host specs so a modest box isn't oversubscribed.
final class VMResourcePlanTests: XCTestCase {
    private let gib = VMResourcePlan.gib

    private func host(cpus: Int, ramGiB: UInt64) -> HostResources {
        HostResources(cpuCount: cpus, memoryBytes: ramGiB * gib)
    }

    // MARK: - Per-VM memory

    func testComfortableHostGetsTheDefaultVMMemory() {
        // 16 GiB host: 12 GiB spare over the reserve → the 6 GiB default.
        XCTAssertEqual(VMResourcePlan.vmMemoryBytes(host: host(cpus: 8, ramGiB: 16)), 6 * gib)
    }

    func testModestHostClampsDownToWhatItCanSpare() {
        // 8 GiB host: 4 GiB spare → clamped between floor (4) and default (6).
        XCTAssertEqual(VMResourcePlan.vmMemoryBytes(host: host(cpus: 4, ramGiB: 8)), 4 * gib)
        // 9 GiB host: 5 GiB spare → 5 GiB.
        XCTAssertEqual(VMResourcePlan.vmMemoryBytes(host: host(cpus: 4, ramGiB: 9)), 5 * gib)
    }

    func testTinyHostNeverGoesBelowTheViabilityFloor() {
        // The stack isn't viable below 4 GiB — the floor holds even when the
        // host can't spare it (maxVMCount is what says "no" on such a host).
        XCTAssertEqual(VMResourcePlan.vmMemoryBytes(host: host(cpus: 2, ramGiB: 4)), 4 * gib)
        XCTAssertEqual(VMResourcePlan.vmMemoryBytes(host: host(cpus: 2, ramGiB: 2)), 4 * gib)
    }

    // MARK: - Per-VM CPUs

    func testCPUCountLeavesTwoHostCoresAndCapsAtFour() {
        XCTAssertEqual(VMResourcePlan.vmCPUCount(host: host(cpus: 12, ramGiB: 32)), 4)
        XCTAssertEqual(VMResourcePlan.vmCPUCount(host: host(cpus: 8, ramGiB: 16)), 4)
        XCTAssertEqual(VMResourcePlan.vmCPUCount(host: host(cpus: 5, ramGiB: 16)), 3)
        XCTAssertEqual(VMResourcePlan.vmCPUCount(host: host(cpus: 4, ramGiB: 16)), 2)
    }

    func testCPUCountNeverExceedsTheHost() {
        XCTAssertEqual(VMResourcePlan.vmCPUCount(host: host(cpus: 2, ramGiB: 8)), 2)
        XCTAssertEqual(VMResourcePlan.vmCPUCount(host: host(cpus: 1, ramGiB: 8)), 1)
    }

    // MARK: - VM cap

    func testHostTooSmallForTheStackHostsZeroVMs() {
        XCTAssertEqual(VMResourcePlan.maxVMCount(host: host(cpus: 2, ramGiB: 4)), 0)
        XCTAssertEqual(VMResourcePlan.maxVMCount(host: host(cpus: 2, ramGiB: 7)), 0)
    }

    func testModestHostHostsExactlyOne() {
        // 8 GiB: one VM at the clamped floor — never oversubscribed to two.
        XCTAssertEqual(VMResourcePlan.maxVMCount(host: host(cpus: 4, ramGiB: 8)), 1)
        // 12 GiB: 8 GiB spare fits one comfortable VM, not two.
        XCTAssertEqual(VMResourcePlan.maxVMCount(host: host(cpus: 8, ramGiB: 12)), 1)
    }

    func testBiggerHostsScaleByTheComfortableDefault() {
        XCTAssertEqual(VMResourcePlan.maxVMCount(host: host(cpus: 8, ramGiB: 16)), 2)  // 12/6
        XCTAssertEqual(VMResourcePlan.maxVMCount(host: host(cpus: 10, ramGiB: 32)), 4) // 28/6
    }

    func testAbsoluteCeilingHolds() {
        XCTAssertEqual(VMResourcePlan.maxVMCount(host: host(cpus: 32, ramGiB: 128)), 8)
    }
}

File: apps/burner-windows/tests/VMResourcePlanTests.cs
using Xunit;
using Flagship.Burner.VM;

namespace Flagship.Burner.Tests;

/// <summary>
/// Direct port of apps/burner-mac VMResourcePlanTests.swift: locks the
/// resource-cap math from docs/desktop-vm-appliance.md. Also pinned by the
/// shared vectors (VMCoreVectorTests).
/// </summary>
public class VMResourcePlanTests
{
    private const ulong GiB = VMResourcePlan.GiB;

    private static HostResources Host(int cpus, ulong ramGiB) => new(cpus, ramGiB * GiB);

    // ---- Per-VM memory ----

    [Fact]
    public void ComfortableHostGetsTheDefaultVMMemory()
    {
        // 16 GiB host: 12 GiB spare over the reserve → the 6 GiB default.
        Assert.Equal(6 * GiB, VMResourcePlan.VmMemoryBytes(Host(8, 16)));
    }

    [Fact]
    public void ModestHostClampsDownToWhatItCanSpare()
    {
        // 8 GiB host: 4 GiB spare → clamped between floor (4) and default (6).
        Assert.Equal(4 * GiB, VMResourcePlan.VmMemoryBytes(Host(4, 8)));
        // 9 GiB host: 5 GiB spare → 5 GiB.
        Assert.Equal(5 * GiB, VMResourcePlan.VmMemoryBytes(Host(4, 9)));
    }

    [Fact]
    public void TinyHostNeverGoesBelowTheViabilityFloor()
    {
        // The stack isn't viable below 4 GiB — the floor holds even when the
        // host can't spare it (MaxVMCount is what says "no" on such a host).
        Assert.Equal(4 * GiB, VMResourcePlan.VmMemoryBytes(Host(2, 4)));
        Assert.Equal(4 * GiB, VMResourcePlan.VmMemoryBytes(Host(2, 2)));
    }

    // ---- Per-VM CPUs ----

    [Fact]
    public void CpuCountLeavesTwoHostCoresAndCapsAtFour()
    {
        Assert.Equal(4, VMResourcePlan.VmCpuCount(Host(12, 32)));
        Assert.Equal(4, VMResourcePlan.VmCpuCount(Host(8, 16)));
        Assert.Equal(3, VMResourcePlan.VmCpuCount(Host(5, 16)));
        Assert.Equal(2, VMResourcePlan.VmCpuCount(Host(4, 16)));
    }

    [Fact]
    public void CpuCountNeverExceedsTheHost()
    {
        Assert.Equal(2, VMResourcePlan.VmCpuCount(Host(2, 8)));
        Assert.Equal(1, VMResourcePlan.VmCpuCount(Host(1, 8)));
    }

    // ---- VM cap ----

    [Fact]
    public void HostTooSmallForTheStackHostsZeroVMs()
    {
        Assert.Equal(0, VMResourcePlan.MaxVMCount(Host(2, 4)));
        Assert.Equal(0, VMResourcePlan.MaxVMCount(Host(2, 7)));
    }

    [Fact]
    public void ModestHostHostsExactlyOne()
    {
        // 8 GiB: one VM at the clamped floor — never oversubscribed to two.
        Assert.Equal(1, VMResourcePlan.MaxVMCount(Host(4, 8)));
        // 12 GiB: 8 GiB spare fits one comfortable VM, not two.
        Assert.Equal(1, VMResourcePlan.MaxVMCount(Host(8, 12)));
    }

    [Fact]
    public void BiggerHostsScaleByTheComfortableDefault()
    {
        Assert.Equal(2, VMResourcePlan.MaxVMCount(Host(8, 16)));   // 12/6
        Assert.Equal(4, VMResourcePlan.MaxVMCount(Host(10, 32)));  // 28/6
    }

    [Fact]
    public void AbsoluteCeilingHolds()
    {
        Assert.Equal(8, VMResourcePlan.MaxVMCount(Host(32, 128)));
    }
}
